Add equalsApprox to vec2, vec3, vec4 (and tests)

querielo opened the pull request:
The PR adds equalsApprox to vec2, vec3, vec4 and fixes the 4 point of [this review comment](https://github.com/playcanvas/engine/pull/5145#issuecomment-1465164912)

> No other math class (like Vec3) offers the capability either

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

Maksims commented:
"Float math is great".
Is it worth checking first if values are actually equal so to early return true, and if not, then checking if their abs difference are within epsilon?

querielo commented:
@Maksims, it depends on real-world vectors. Probably, if the vectors are equal and not computed, we will get a performance boost. However, if you use the method, that means you computed two vectors and equals may not be suitable for your situation.

Here is a small performance test: 
https://jsbench.me/jvlmu8cc2d/1

querielo commented:
It seems that we will only achieve a performance boost if there are too many identical vectors.

On my computer, the times are equal for the following vectors
```
let unique = [
	... new Array(30).fill(new Vec4(1,2,3,4)),
	new Vec4(1,2,3,4 + 0.0000001),
	new Vec4(1,2,3,4 - 0.0000001),
	new Vec4(1,2,3,4 + 0.00001),
	new Vec4(1,2,3,4 - 0.00001),
	new Vec4(1 - 0.0000001,2,3,4),
	new Vec4(1 + 0.0000001,2,3,4),
	new Vec4(1 - 0.00001,2,3,4),
	new Vec4(1 + 0.00001,2,3,4),
]
```

Link: https://jsbench.me/jvlmu8cc2d/2

mvaligursky reviewed (approved):
Nice work, thanks!

willeastcott commented:
What about `Color`?

mvaligursky commented:
You don't typically do math on Colors much .. that's why our Color class is so lite with almost no math functionality.

willeastcott commented:
OK, fair enough.

Also, as folks probably know, I'm always hesitant about bloating the API unnecessarily. Any reason why this is not sufficient:

```
if (a.distance(b) < epsilon) {
    console.log('approximately equal');
}
```

kungfooman commented:
> Also, as folks probably know, I'm always hesitant about bloating the API unnecessarily. Any reason why this is not sufficient:

I have the same objection, that's why I usually try to find cases in the engine that could be shortened/simplified through a PR in the engine itself to prove the "real world need".

The problem with `Vec3#distance` is usually that people say it's slooow because of the `Math.sqrt` and then people use `Vec3#lengthSq` instead while multiplying the epsilon with itself to make up for the missing square root:

https://github.com/playcanvas/engine/blob/2bfd57bac9a7f31c1306ecb13541784ce1075027/src/framework/input/element-input.js#L101-L106

(maybe that's at least one case which could be simplified with this PR and maybe there are others)

LeXXik commented:
If I rarely need an approximate comparison, I go with Will's example. Not sure I'd expect it to be part of API.

mvaligursky commented:
We already have this function for Quat, so it's nice to be consistent. In the past I used those function for vectors in different engines.

querielo commented:
Added 2 tests: https://jsbench.me/jvlmu8cc2d/3

The 4th test is the @willeastcott's [suggestion](https://github.com/playcanvas/engine/pull/5662#issuecomment-1731120899)
The 5th test is the @kungfooman's [suggestion](https://github.com/playcanvas/engine/pull/5662#issuecomment-1731153629)

Please, check if I implemented them correctly. 

Note: there is no distance function in vec4.


My result:
<img width="794" alt="Screenshot 2023-09-22 at 12 32 37" src="https://github.com/playcanvas/engine/assets/104348270/b53a5b38-4fcf-43ca-ab2e-54778acd4a10">

querielo commented:
Looks like in edge cases the original equalApprox provides the stablest result.

* Many different vectors in the first coordinate
<img width="813" alt="Screenshot 2023-09-22 at 12 43 53" src="https://github.com/playcanvas/engine/assets/104348270/50883c32-176b-4b06-b27b-8fb09fa012df">
* Many different vectors in the last coordinate
<img width="878" alt="Screenshot 2023-09-22 at 12 47 49" src="https://github.com/playcanvas/engine/assets/104348270/0b4b6e06-c3c1-4c89-8d37-ca10f3470eab">
* Many equal vectors
<img width="878" alt="Screenshot 2023-09-22 at 12 51 26" src="https://github.com/playcanvas/engine/assets/104348270/6dc7b5b2-55dc-4f0b-999a-ee19e5339383">
